client/pd_service_discovery.go

```diff
@@ -1150,15 +1121,28 @@ func (c *pdServiceDiscovery) GetOrCreateGRPCConn(addr string) (*grpc.ClientConn,
 	return grpcutil.GetOrCreateGRPCConn(c.ctx, &c.clientConns, addr, c.tlsCfg, c.option.gRPCDialOptions...)
 }
 
-func addrsToUrls(addrs []string) []string {
+func addrsToUrls(addrs []string, tlsCfg *tls.Config) []string {
 	// Add default schema "http://" to addrs.
 	urls := make([]string, 0, len(addrs))
 	for _, addr := range addrs {
-		if strings.Contains(addr, "://") {
-			urls = append(urls, addr)
-		} else {
-			urls = append(urls, "http://"+addr)
-		}
+		urls = append(urls, addrToUrl(addr, tlsCfg))
 	}
 	return urls
 }
+
+func addrToUrl(addr string, tlsCfg *tls.Config) string {
+	if tlsCfg == nil {
+		if strings.HasPrefix(addr, httpsScheme) {
+			addr = fmt.Sprintf("%s%s", httpScheme, strings.TrimPrefix(addr, httpsScheme))
+		} else if !strings.HasPrefix(addr, httpScheme) {
+			addr = fmt.Sprintf("%s%s", httpScheme, addr)
+		}
+	} else {
+		if strings.HasPrefix(addr, httpScheme) {
```

Comment:
Shall we according to the etcd member URL change the prefix instead of tlsCfg?

Reply:
This is only used for `url.Parse`. If we don't add scheme, it will return error.
If etcd member URL is not same as this, the connection map will add a new KV pair.